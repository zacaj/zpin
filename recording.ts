import { Log } from './log';
import { Timer, passTime, setTime, time, wait } from './timer';
import * as fs from 'fs';
import { split, assert, debugging } from './util';
import { matrix } from './switch-matrix';
import { settleForks } from './promises';
import { MPU } from './mpu';
import { initMachine } from './init';
import { Events } from './events';
import { machine } from './machine';

export let curRecording: string|undefined = undefined;
let lines: string[];
let curTime = 0;
let curLine = 0;

export function initRecording(recording: string) {
    curRecording = recording;
    Log.log(['console'], 'playing back recording %s', recording);
    Timer.mockTime = 0;
    const file = fs.readFileSync(curRecording!, 'utf8');
    lines = file.split('\n');
    curTime = 0;
    curLine = 1;
}

export async function playRecording(toPoint?: string) {
    // await new Promise(r => setTimeout(r, 10000));
    const events = Events;
    const timer = Timer;
    const _machine = machine;
    for (; curLine < lines.length; curLine++) {
        const line = lines[curLine];
        if (line.startsWith('#')) continue;
        if (line && !line.includes(' ')) {
            const bp = line;
            if (bp === 'debug' && debugging()) {
                const nextLine = lines[curLine+1];
                debugger;
            } else if (process.env.NODE_ENV === 'test') {
                assert(bp === toPoint, `expected breakpoint ${toPoint} but got '${bp}'`);
                curLine++;
                await settleForks();
                return;
            }
            continue;
        }
        const [ts, sw, stateS, colS, rowS, timeS] = split(line, ' switch \'', '\' state -> ', ' (', ',', ') at ');
        if (timeS === undefined) continue;

        const time = parseInt(timeS, 10);
        const diff = time - curTime;
        assert(diff>=0, 'time travel not implemented');
        if (diff)
            await passTime(diff);
        curTime = time;

        matrix[parseInt(colS, 10)][parseInt(rowS, 10)]!.changeState(stateS === 'true', 'recording '+curRecording);
        await settleForks();
    }
    await settleForks();

    if (process.env.NODE_ENV !== 'test')
        await setTime(undefined);
    Log.log(['console', 'switch'], 'recording %s finished at %i', curRecording, time());
    curRecording = undefined;
}


export async function testRecording(name: string, toPoint?: string) {
    jest.spyOn(Log, 'init').mockImplementation(() => {});
    jest.spyOn(MPU, 'sendCommandCode').mockImplementation(async (cmd) => {
        return {
            code: 200,
            resp: 'mocked',
        };
    });
    await initMachine(false, false, true, false, `./recordings/${name}.rec`, toPoint);
}

export async function continueRecording(toPoint?: string) {
    await playRecording(toPoint);
}
export async function finishRecording() {
    await playRecording();
}
