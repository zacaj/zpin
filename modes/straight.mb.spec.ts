import { machine } from '../machine';
import { testRecording, finishRecording } from '../recording';

describe('straight mb', () => {
    test('ramp lowers', async () => {
        await testRecording('lightStraight');
        expect(machine.cRamp.val).toBe(false);
    });
});