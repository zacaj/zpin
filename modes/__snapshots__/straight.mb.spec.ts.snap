// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`straight mb lock holds ball 1`] = `
Object {
  "outputs": Object {
    "centerBank": false,
    "getSkillshot": [Function],
    "iCenter1": "redArrow",
    "iCenter2": "",
    "iCenter3": "",
    "iLeft1": "",
    "iLeft2": "",
    "iLeft3": "greenArrow",
    "iLeft4": "greenArrow",
    "iMini1": "",
    "iMini2": "",
    "iMini3": "",
    "iRight1": "",
    "iRight2": "",
    "iRight3": "",
    "iRight4": "yellowArrow",
    "iRight5": "yellowArrow",
    "iSS1": Object {
      "fill": [Function],
    },
    "iSS2": Object {
      "fill": [Function],
    },
    "iSS3": Object {
      "fill": [Function],
    },
    "iSS4": Object {
      "fill": [Function],
    },
    "iSS5": Object {
      "fill": [Function],
    },
    "iSS6": Object {
      "fill": [Function],
    },
    "iSS7": Object {
      "fill": [Function],
    },
    "iSpinner": "",
    "iUpper21": "orangeArrow",
    "iUpper22": "orangeArrow",
    "iUpper31": "blueArrow",
    "iUpper32": "blueArrow",
    "iUpper33": "blueArrow",
    "kickerEnable": true,
    "lEjectShowCards": Array [],
    "lEjectStartMode": Array [],
    "lLowerRamp": Array [],
    "lMagnaSaveStatus": Array [
      "Green",
    ],
    "lMiniReady": Array [
      "Red",
    ],
    "lPopperStatus": Array [
      "Green",
    ],
    "lPower1": Array [
      "Orange",
    ],
    "lPower2": Array [],
    "lPower3": Array [],
    "lPower4": Array [],
    "lRampArrow": Array [],
    "lRampShowCards": Array [],
    "lRampStartMb": Array [],
    "lShooterShowCards": Array [],
    "lShooterStartHand": Array [],
    "leftBank": false,
    "leftGate": false,
    "leftMagnet": false,
    "lockPost": false,
    "magnetPost": false,
    "miniBank": false,
    "miniDiverter": false,
    "miniEject": false,
    "miniFlipperEnable": false,
    "outhole": false,
    "popper": false,
    "rampUp": true,
    "realRightBank": false,
    "right1": false,
    "right2": false,
    "right3": false,
    "right4": false,
    "right5": false,
    "rightBank": false,
    "rightGate": false,
    "shooterDiverter": true,
    "troughRelease": false,
    "upper2": false,
    "upper3": false,
    "upperEject": false,
    "upperMagnet": false,
  },
  "state": Object {
    "children": Array [
      Object {
        "children": Array [],
        "name": "DropBank46",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank47",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank48",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank49",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank50",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank51",
        "state": Object {},
      },
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "name": "EosPulse53",
            "state": Object {},
          },
        ],
        "name": "MachineOverrides52",
        "state": Object {},
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "name": "ClearHoles56",
                "state": Object {},
              },
              Object {
                "children": Array [],
                "name": "Spinner58",
                "state": Object {
                  "comboMult": 1,
                  "rounds": 0,
                  "score": 300,
                },
              },
              Object {
                "children": Array [],
                "name": "PlayerOverrides59",
                "state": Object {},
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter62",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter63",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter64",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter65",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter66",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter67",
                        "state": Object {},
                      },
                    ],
                    "name": "ResetAnyDropOnComplete61",
                    "state": Object {},
                  },
                  Object {
                    "children": Array [],
                    "name": "StraightMb71",
                    "state": Object {
                      "awardingJp": 0,
                      "curBank": undefined,
                      "jackpotLit": false,
                      "lockPost": false,
                    },
                  },
                  Object {
                    "children": Array [],
                    "name": "Skillshot72",
                    "state": Object {
                      "curAward": 0,
                      "shooterOpen": true,
                    },
                  },
                ],
                "name": "Ball60",
                "state": Object {
                  "miniReady": false,
                },
              },
            ],
            "name": "Player55",
            "state": Object {
              "chips": 1,
              "closeShooter": false,
              "curMbMode": "StraightMb71",
              "lowerRampLit": false,
              "mbsQualified": Array [],
              "modesQualified": Array [
                13,
              ],
              "poker": undefined,
              "rampUp": true,
              "score": 40000,
            },
          },
        ],
        "name": "Game54",
        "state": Object {
          "ballNum": 1,
        },
      },
    ],
    "name": "Machine45",
    "state": Object {
      "lastSwitchHit": Switch {
        "_state": false,
        "column": 0,
        "lastChange": 53612,
        "lastClosed": 48957,
        "lastOpened": 53612,
        "minOffTime": 10,
        "minOnTime": 1,
        "name": "ramp made",
        "row": 7,
      },
      "lockDown": true,
      "miniDown": false,
    },
  },
}
`;

exports[`straight mb ramp lowers 1`] = `
Object {
  "outputs": Object {
    "centerBank": false,
    "getSkillshot": [Function],
    "iCenter1": "",
    "iCenter2": "",
    "iCenter3": "",
    "iLeft1": "",
    "iLeft2": "",
    "iLeft3": "",
    "iLeft4": "",
    "iMini1": "",
    "iMini2": "",
    "iMini3": "",
    "iRight1": "",
    "iRight2": "",
    "iRight3": "",
    "iRight4": "",
    "iRight5": "",
    "iSS1": "",
    "iSS2": "",
    "iSS3": "",
    "iSS4": "",
    "iSS5": "",
    "iSS6": "",
    "iSS7": "",
    "iSpinner": "",
    "iUpper21": "",
    "iUpper22": "",
    "iUpper31": "",
    "iUpper32": "",
    "iUpper33": "",
    "kickerEnable": true,
    "lEjectShowCards": Array [
      "Green",
    ],
    "lEjectStartMode": Array [
      "White",
    ],
    "lLowerRamp": Array [],
    "lMagnaSaveStatus": Array [
      "Green",
    ],
    "lMiniReady": Array [
      "Red",
    ],
    "lPopperStatus": Array [
      "Green",
    ],
    "lPower1": Array [
      "Orange",
    ],
    "lPower2": Array [],
    "lPower3": Array [],
    "lPower4": Array [],
    "lRampArrow": Array [],
    "lRampShowCards": Array [
      "Green",
    ],
    "lRampStartMb": Array [
      "White",
    ],
    "lShooterShowCards": Array [
      "Green",
    ],
    "lShooterStartHand": Array [
      "White",
    ],
    "leftBank": false,
    "leftGate": false,
    "leftMagnet": false,
    "lockPost": false,
    "magnetPost": false,
    "miniBank": false,
    "miniDiverter": false,
    "miniEject": false,
    "miniFlipperEnable": false,
    "outhole": false,
    "popper": false,
    "rampUp": false,
    "realRightBank": false,
    "right1": false,
    "right2": false,
    "right3": false,
    "right4": false,
    "right5": false,
    "rightBank": false,
    "rightGate": false,
    "shooterDiverter": true,
    "troughRelease": false,
    "upper2": false,
    "upper3": false,
    "upperEject": false,
    "upperMagnet": false,
  },
  "state": Object {
    "children": Array [
      Object {
        "children": Array [],
        "name": "DropBank11",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank12",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank13",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank14",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank15",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank16",
        "state": Object {},
      },
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "name": "EosPulse18",
            "state": Object {},
          },
        ],
        "name": "MachineOverrides17",
        "state": Object {},
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "name": "ClearHoles21",
                "state": Object {},
              },
              Object {
                "children": Array [],
                "name": "Poker22",
                "state": Object {
                  "bet": 100,
                  "closeShooter": false,
                  "dealerCardsUsed": Array [],
                  "dealerHand": Array [
                    Object {
                      "flipped": true,
                      "num": 3,
                      "suit": "h",
                    },
                    Object {
                      "flipped": true,
                      "num": 4,
                      "suit": "c",
                    },
                    Object {
                      "flipped": false,
                      "num": 9,
                      "suit": "h",
                    },
                    Object {
                      "flipped": false,
                      "num": 5,
                      "suit": "s",
                    },
                    Object {
                      "flipped": false,
                      "num": 9,
                      "suit": "d",
                    },
                    Object {
                      "flipped": false,
                      "num": 4,
                      "suit": "h",
                    },
                    Object {
                      "flipped": true,
                      "num": 2,
                      "suit": "c",
                    },
                  ],
                  "dealerHandDesc": "Pair",
                  "newMbs": Array [
                    Array [
                      "StraightMb",
                      Array [
                        Object {
                          "num": 7,
                          "suit": "s",
                        },
                        Object {
                          "num": 8,
                          "suit": "c",
                        },
                        Object {
                          "num": 9,
                          "suit": "c",
                        },
                        Object {
                          "num": 10,
                          "suit": "s",
                        },
                        Object {
                          "num": 11,
                          "suit": "s",
                        },
                      ],
                    ],
                  ],
                  "newModes": Array [
                    13,
                  ],
                  "playerCardsUsed": Array [],
                  "playerHand": Array [
                    Object {
                      "num": 7,
                      "suit": "s",
                    },
                    Object {
                      "num": 11,
                      "suit": "s",
                    },
                    Object {
                      "num": 8,
                      "suit": "c",
                    },
                    Object {
                      "num": 9,
                      "suit": "c",
                    },
                    Object {
                      "num": 10,
                      "suit": "s",
                    },
                    Object {
                      "num": 13,
                      "suit": "s",
                    },
                    Object {
                      "num": 13,
                      "suit": "c",
                    },
                  ],
                  "playerHandDesc": "Straight",
                  "playerWins": undefined,
                  "pot": 1000,
                  "slots": Array [
                    Object {
                      "num": 1,
                      "suit": "d",
                    },
                    null,
                    null,
                    null,
                    Object {
                      "num": 1,
                      "suit": "s",
                    },
                    Object {
                      "num": 13,
                      "suit": "h",
                    },
                    Object {
                      "num": 3,
                      "suit": "d",
                    },
                    Object {
                      "num": 13,
                      "suit": "d",
                    },
                    null,
                    null,
                    Object {
                      "num": 11,
                      "suit": "h",
                    },
                    Object {
                      "num": 8,
                      "suit": "s",
                    },
                    Object {
                      "num": 10,
                      "suit": "h",
                    },
                    Object {
                      "num": 12,
                      "suit": "d",
                    },
                    Object {
                      "num": 8,
                      "suit": "d",
                    },
                    Object {
                      "num": 10,
                      "suit": "d",
                    },
                    Object {
                      "num": 12,
                      "suit": "s",
                    },
                    Object {
                      "num": 1,
                      "suit": "h",
                    },
                    Object {
                      "num": 9,
                      "suit": "s",
                    },
                    Object {
                      "num": 1,
                      "suit": "c",
                    },
                  ],
                  "step": 7,
                },
              },
              Object {
                "children": Array [],
                "name": "Spinner23",
                "state": Object {
                  "comboMult": 1,
                  "rounds": 0,
                  "score": 300,
                },
              },
              Object {
                "children": Array [],
                "name": "PlayerOverrides24",
                "state": Object {},
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter27",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter28",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter29",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter30",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter31",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter32",
                        "state": Object {},
                      },
                    ],
                    "name": "ResetAnyDropOnComplete26",
                    "state": Object {},
                  },
                ],
                "name": "Ball25",
                "state": Object {
                  "miniReady": false,
                },
              },
            ],
            "name": "Player20",
            "state": Object {
              "chips": 1,
              "closeShooter": false,
              "curMbMode": undefined,
              "lowerRampLit": false,
              "mbsQualified": Array [],
              "modesQualified": Array [],
              "poker": "Poker22",
              "rampUp": true,
              "score": 35000,
            },
          },
        ],
        "name": "Game19",
        "state": Object {
          "ballNum": 1,
        },
      },
    ],
    "name": "Machine10",
    "state": Object {
      "lastSwitchHit": Switch {
        "_state": true,
        "column": 3,
        "lastChange": 30053,
        "lastClosed": 30053,
        "lastOpened": undefined,
        "minOffTime": 150,
        "minOnTime": 10,
        "name": "right 3",
        "row": 2,
      },
      "lockDown": false,
      "miniDown": false,
    },
  },
}
`;
